### PR TITLE
chore: codebase quality sweep — bug fixes, dead-code removal, file org

### PR DESCRIPTION
**Pull request by nicholasjconn**
## Summary

Autonomous `/improve-codebase` sweep focused on `config` / `clients` / `warnings` / `tools` / `publish-site` / `upgrade-migrations`. The pass fixes two real bugs, removes zero-caller dead code, reorganizes one oversized file with no behavior change, and tightens determinism and fail-loud behavior. All local checks are green (`make ci`: tidy-check, test-release, coverage 95.34% ≥ 95%, online e2e with required upgrade scenarios, docs-cta-check).

## Highlights

- **Release-critical fix — `extractchecksum` `TrimLeft` → `TrimPrefix`.** `strings.TrimLeft(name, "./")` stripped `.` and `/` as a cutset, mangling dotfile basenames (e.g. `.hidden-file.bin` lost its leading dot); `TrimPrefix` strips only the `./` path marker. Regression guard added in `scripts/test-release/tool_tests.sh`.
- **`publish-site` H2 slug-collision fix.** Colliding H2 headings now get de-duplicated anchors (`foo`, `foo-1`, `foo-2`, …) matching Docusaurus's github-slugger output, including the suffix-collision edge case where a natural slug clashes with a generated suffix (`foo`, `foo-1`, `foo-1-1`).
- **Zero-caller dead-code removal in `internal/config`.** Deleted `commands.go`, `instructions.go` and their tests; removed the unused `LoadConfig` loader. Production paths use the `*FS` variants; `utf8BOM`/`osReadFileFunc` moved to `skills.go`.
- **`upgrade_migrations` file split (no behavior change).** Skills-migration code extracted from `upgrade_migrations.go` into `upgrade_migrations_skills.go` for readability.
- **`skills.go` fail-loud duplicate-key check.** Reject duplicate front-matter `metadata` keys (parity with top-level keys) instead of silently letting the last value win.
- **Determinism / test improvements.** Exact-value assertions in `warnings` token estimation, expanded `version.IsDev` and probe `commandExitCode` branch coverage, exact `vscode` path-set assertions, best-effort stderr warning writes in the `claude`/`codex` launchers.

## Deferred (see `docs/agent-layer/ISSUES.md`)

- **secret-in-URL precision/recall decision — reverted pending a human call.** An `/improve-codebase` rewrite of `looksLikeSecretQueryKey` (substring vs. word-segment matching) was reverted because it is a genuine security precision/recall tradeoff, not a clear win. Logged as `secret-in-url-precision-recall` for the maintainer to decide.
- Architectural items logged for a focused follow-up: `fs-os-abstraction-four-patterns` (one OS/filesystem testability seam implemented four ways) and `semver-parse-compare-duplicated-three-packages` (semver parse/compare duplicated across install/update/dispatch).

## Note

`docs/agent-layer/BACKLOG.md` contains an **unrelated, pre-existing user-authored backlog entry** (`interactive-html-review-skill`) that was already uncommitted before this sweep began. It rides along in this commit as a benign memory-file addition and is not part of the sweep.

## Release Notes

* **New Features**
  * Added a skills-format migration workflow to convert flat skill files into directory structure.

* **Improvements**
  * Enhanced interactive `al upgrade` prompts and dry-run plan/summary rendering with standardized messaging.
  * Centralized semantic version comparison/parsing for upgrade and migration planning.
  * Site publishing now generates Docusaurus heading anchors with robust collision handling.

* **Bug Fixes**
  * Upgrade/migration flows now use consistent, standardized messaging and improved output rendering.
  * Improved checksum extraction handling for `./` targets and dotfile names.

* **Tests**
  * Added collision-resolution and tooling regression coverage; expanded upgrade readiness/migration test assertions.

### CODE REVIEW

File: cmd/publish-site/main.go [hunk 1]
```diff
@@ -478,10 +488,23 @@ func collectLevelTwoHeadings(body string) []guideHeading {
...
+		if n > 0 {
+			slug = fmt.Sprintf("%s-%d", baseSlug, n)
+		}
+		// If the candidate slug was already emitted (suffix-collision with a
+		// heading whose natural slug matches), keep incrementing.
+		for emittedSlugs[slug] {
+			n++
+			slug = fmt.Sprintf("%s-%d", baseSlug, n)
+		}
+		slugCounts[baseSlug]++
```
[the remark this answers is not in the document]
Comment (nicholasjconn) [reply cmd/publish-site/main.go]: **Fixed in `05eb26e`.** Adopted the suggestion — `collectLevelTwoHeadings` now sets `slugCounts[baseSlug] = n + 1` instead of `++`, so the next repeated heading starts searching from the suffix already used rather than re-walking claimed low suffixes. This avoids the quadratic worst case you flagged. Output is unchanged: the `emittedSlugs` set still guarantees uniqueness, and this matches github-slugger's behavior (which advances its per-base counter inside the same collision loop). Verified by `TestCollectLevelTwoHeadings_DeduplicatesCollidingSlugs` and `TestCollectLevelTwoHeadings_SuffixCollisionResolved`.

**Comment by nicholasjconn** [comment]
**Fixed in `05eb26e`.** Re: the automated reviews —

- **gemini-code-assist:** the one inline finding (potential quadratic complexity in the H2 slug de-duplication loop) is addressed in `05eb26e` — the base counter now advances to `n + 1` past the suffix actually used, eliminating the re-walk of claimed low suffixes while preserving output and uniqueness. Replied inline and resolved that thread.
- **Copilot:** reviewed all 30 files and generated no comments; no change required.
- **CodeRabbit:** hit its per-developer review rate limit and did not produce findings; nothing to address.

CI (`verify` + `website-build`) is green on the latest commit.

**Comment by nicholasjconn** [comment]
**Fixed in `943bcdf`.**

CodeRabbit's nitpick on `internal/clients/codex/launch_test.go` lines 106–108 — replace the fixed `[4096]byte` buffer + single `Read` call with `io.ReadAll`: applied. The test now uses `io.ReadAll(r)` and checks the read error before converting to string. More idiomatic, handles multi-part writes, and provides better error visibility. Verified with `go test ./internal/clients/codex/...` (all pass) and `make lint` (0 issues).
